### PR TITLE
Adding monaco editor to code mode

### PR DESCRIPTION
**Pull request by habdelra**
This PR adds the monaco editor to code mode (along with the monaco editor auto saving). I wasn't too sure where to put the auto saving indicator (we show this in the code route). Probably we'll need Gardella to design that and we can add it in a followup PR.

The primary difference in percy for this PR is that we are no longer rendering an extra page beyond the last line--it's an option in the monaco API that i disabled since it looks awkward being able to scroll thru a whole page of emptiness in the code editor. (also the way percy renders monaco is really awkward--lol)

monaco-code-mode

### CODE REVIEW

File: packages/host/app/services/monaco-service.ts [hunk 1]
```diff
@@ -38,6 +42,39 @@ export default class MonacoService extends Service {
...
+  //
+  // In order to prevent confusion around which type of serialization you are dealing
+  // with, we convert the file serialization back to the form the host is accustomed
+  // to (application/vnd.card+json) as soon as possible so that the semantics around
+  // file serialization don't leak outside of where they are immediately used.
+  reverseFileSerialization(
+    fileSerializationJSON: SingleCardDocument,
+    id: string,
+  ): SingleCardDocument {
+    let realmURL = this.cardService.getRealmURLFor(new URL(id))?.href;
```
Comment (jurgenwerk) [inline packages/host/app/services/monaco-service.ts]: I think I see a case here where id is the url of a card that belongs to an unknown realm. In that case, the code below will reach the error state of "Could not determine realm for url".

Maybe you can pass an additional `realmURL` argument to this function, which you can get from `this.openFile.current.realmURL` in code mode component. That value is extracted from the boxel realm header.

**Review by jurgenwerk — APPROVED** [review]
Looks good but left a comment about a potential edge case of when the card belongs to an unknown realm

**Review by FadhlanR — APPROVED** [review]

File: packages/host/app/services/monaco-service.ts (hunk 1 above)
Comment (habdelra) [reply packages/host/app/services/monaco-service.ts]: fixed in https://github.com/cardstack/boxel/pull/604/commits/7f90c3f355fb48a0d93868cb907b6b55e41bd3d5 thanks!

**Review by burieberry — COMMENTED** [review]
Looks good, but I couldn't get it to scroll. 

Also - may not be within the scope of this PR - after making a change in the editor and switching back to `interact` mode, only after a page refresh I see the change.

**Comment by habdelra** [comment]
> Looks good, but I couldn't get it to scroll.
> 
> Also - may not be within the scope of this PR - after making a change in the editor and switching back to `interact` mode, only after a page refresh I see the change.

for the scroll, could you post a screenshot? did you have more than one page of content?

regarding the switch to interact mode, the cards in interact mode are currently not live refreshing. probably we'll need to make another PR for that. I can file a ticket around that after this is merged

**Comment by habdelra** [comment]
When I have more than one page of content I am seeing scrolling just fine 
scroll-monaco

**Comment by habdelra** [comment]
Note that I preemptively filed this ticket around live updating cards in interact mode: https://linear.app/cardstack/issue/CS-5963/bug-cards-that-are-modified-in-code-mode-do-not-have-their-changes

**Comment by burieberry** [comment]
> When I have more than one page of content I am seeing scrolling just fine scroll-monaco

No I just made the screen small. If we don't worry about that, it's fine then

**Review by burieberry — APPROVED** [review]

**Comment by habdelra** [comment]
> No I just made the screen small. If we don't worry about that, it's fine then

in the latest commit i turned on a property for auto resize to accommodate the new resizing panels that I think actually fixed the issue for small screens too.

**Pull request merged**